From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jonathan Leitschuh <jlleitschuh@wpi.edu>
Date: Tue, 19 Aug 2014 15:16:55 -0400
Subject: [PATCH 0307/6262] Fixes debugging in Java

When deploying a Java program in debug mode any previous running programs should be properly removed. [artf3408]
This should link with (https://usfirst.collab.net/sf/go/artf3408?nav=1&_pagenum=2&returnUrlKey=1408475631056)

[artf3415] is still an issue until NI resolves it.
This should link with (https://usfirst.collab.net/sf/go/artf3415?nav=1&_pagenum=1&returnUrlKey=1408475758258)

Change-Id: Ica9639b5a406b70be4d51aa4e45ffef56baddc93
---
 .../resources/java-zip/ant/build.properties   |  5 ++--
 .../src/main/resources/java-zip/ant/build.xml | 24 +++++++++++++------
 .../src/main/resources/java-zip/ant/frcdebug  |  2 ++
 .../resources/java-zip/ant/robotDebugCommand  |  3 +++
 4 files changed, 25 insertions(+), 9 deletions(-)
 create mode 100644 eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/frcdebug
 create mode 100644 eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/robotDebugCommand

diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
index 5a2fd03353637ede3f34693f3e90e0eb78928f21..7c800ce92d30d61f603a2bc7d7c7d7bc5da97f89 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.properties
@@ -2,9 +2,10 @@
 username=admin
 password=
 deploy.dir=/home/admin
-deploy.kill.command=/usr/local/frc/bin/frcKillRobot.sh -t -r
-deploy.debug.command = /usr/local/frc/JRE/bin/java -XX:+UsePerfData -agentlib:jdwp=transport=dt_socket,address=8348,server=y,suspend=y -jar ${deploy.dir}/FRCUserProgram.jar
+deploy.kill.command=. /etc/profile.d/natinst-path.sh; /usr/local/frc/bin/frcKillRobot.sh -t -r
 deploy.log.file=/var/local/natinst/log/FRC_UserProgram.log
+deploy.log.command=tail -F -s 0 -n 0 ${deploy.log.file}
+debug.flag.dir=/tmp/
 command.dir=/home/lvuser/
 
 # Libraries to use
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
index 574a96a05521bdac01b24f54838b7dd31af46101..5a25af19555d472b4748f6b9ac59fa0419f8400e 100644
--- a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/build.xml
@@ -76,25 +76,35 @@
        username="${username}"
        password="${password}"
        trust="true"
-       command=". /etc/profile.d/natinst-path.sh; ${deploy.kill.command};"/>
+       command="${deploy.kill.command};"/>
 
     <sshexec host="${target}"
        username="${username}"
        password="${password}"
        trust="true"
-       command="tail -F -s 0 -n 0 ${deploy.log.file}"/>
+       command="${deploy.log.command}"/>
   </target>
 
   <target name="debug-deploy" depends="get-target-ip,jar" description="Deploy the jar and start the program running.">
     <echo>[athena-deploy] Copying code over.</echo>
     <scp file="${dist.jar}" todir="${username}@${target}:${deploy.dir}" password="${password}" trust="true"/>
+  	<!-- The remoteDebugCommand file is used by /usr/local/frc/bin/frcRunRobot.sh on the roboRIO  -->
+    <scp file="${wpilib.ant.dir}/robotDebugCommand" todir="${username}@${target}:${command.dir}" password="${password}" trust="true"/>
+  	<!-- The frcdebug file is used as a flag for /usr/local/frc/bin/frcRunRobot.sh to run the robot program in debug mode -->
+  	<scp file="${wpilib.ant.dir}/frcdebug" todir="${username}@${target}:${debug.flag.dir}" password="${password}" trust="true"/>
 
-    <echo>[athena-deploy] Starting program.</echo>
+    <echo>[athena-deploy] Starting Debug program.</echo>
     <sshexec host="${target}"
-       username="${username}"
-       password="${password}"
-       trust="true"
-       command=". /etc/profile.d/natinst-path.sh; ${deploy.debug.command}"/>
+        username="${username}"
+        password="${password}"
+        trust="true"
+        command="${deploy.kill.command}"/>
+
+    <sshexec host="${target}"
+        username="${username}"
+        password="${password}"
+        trust="true"
+        command="${deploy.log.command}"/>
   </target>
 
   <!-- Simulate -->
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/frcdebug b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/frcdebug
new file mode 100644
index 0000000000000000000000000000000000000000..206e7f86e816b44493328777c649f6666781fdf4
--- /dev/null
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/frcdebug
@@ -0,0 +1,2 @@
+# This file is used as a flag to determine if debugging should be used.
+# It is uploaded to the robot when launched in debug mode and should be removed automatically once used.
\ No newline at end of file
diff --git a/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/robotDebugCommand b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/robotDebugCommand
new file mode 100644
index 0000000000000000000000000000000000000000..317e4a4c8253aab02c4a8032d10488f730784b65
--- /dev/null
+++ b/eclipse-plugins/edu.wpi.first.wpilib.plugins.java/src/main/resources/java-zip/ant/robotDebugCommand
@@ -0,0 +1,3 @@
+/usr/local/frc/JRE/bin/java -XX:+UsePerfData -agentlib:jdwp=transport=dt_socket,address=8348,server=y,suspend=y
+-jar /home/admin/FRCUserProgram.jar
+
